go/vt/vtgate/planbuilder/horizon_planning.go

```diff
@@ -145,19 +145,61 @@ func (hp *horizonPlanning) planAggregations() error {
 	}
 
 	for _, e := range hp.qp.SelectExprs {
-		offset, _, err := pushProjection(e.Col, hp.plan, hp.semTable, true, false)
-		if err != nil {
-			return err
+		// push all expression if they are non-aggregating or the plan is not ordered aggregated plan.
+		if !e.Aggr || oa == nil {
+			_, _, err := pushProjection(e.Col, hp.plan, hp.semTable, true, false)
+			if err != nil {
+				return err
+			}
+			continue
 		}
+
 		if e.Aggr && oa != nil {
 			fExpr, isFunc := e.Col.Expr.(*sqlparser.FuncExpr)
 			if !isFunc {
 				return vterrors.Errorf(vtrpcpb.Code_UNIMPLEMENTED, "unsupported: in scatter query: complex aggregate expression")
 			}
 			opcode := engine.SupportedAggregates[fExpr.Name.Lowered()]
-			oa.eaggr.Aggregates = append(oa.eaggr.Aggregates, engine.AggregateParams{
+			handleDistinct, innerAliased, err := oa.needDistinctHandlingGen4(fExpr, opcode, hp.semTable, hp.vschema)
+			if err != nil {
+				return err
+			}
+
+			// Currently the OA engine primitive is able to handle only one distinct aggregation function.
+			// PreProcess being true tells that it is already handling it.
+			if oa.eaggr.PreProcess && handleDistinct {
+				return semantics.Gen4NotSupportedF("multiple distinct aggregation function")
```

Comment:
this is a limitation with the engine, and not the planner, if I understand it correctly. `Gen4NotSupportedF` should only be used to mark queries the gen4 planner should handle but doesn't. this is different, IMHO